### PR TITLE
gh-118306: Update JIT compilation to use LLVM 18

### PR DESCRIPTION
**Pull request by savannahostrowski**
This PR updates the LLVM version used for JIT compilation from 16 to 18.

Main changes, other than the basic `16` --> `18` string updates 😊  :
- Handled some new instructions that were previously not surfaced in LLVM 16 across OS binary formats
- Updated properties where JSON parser had changed

Known Issues:
- The `--with-lto` option had to be removed for emulated and native Linux in CI due to an open issue with llvm-18 on Ubuntu 22.04. The issue is that `LLVMgold.so` is currently missing from binaries (see this [failed CI run](https://github.com/savannahostrowski/cpython/actions/runs/8842831500/job/24282099980) as an example). The current workaround folks are using is to build without LTO. See https://github.com/llvm/llvm-project/issues/87553 for more details. Not ideal but @brandtbucher I'll let you be the judge as to whether this is acceptable in the short-term. I've left an inline comment in the workflow yml about this.
- `-mcmodel=large` seems to be unsupported in LLVM 18 so I had to remove this from the build args in `_targets.py`. See https://github.com/llvm/llvm-project/pull/70262 for more details.

* Issue: gh-118306

### CODE REVIEW

**Comment by savannahostrowski** [comment]
It looks like `test_traceback` and `test_tracemalloc` are failing in aarch64 linux jobs. However, those also seem to be failing in other JIT CI runs as well?

**Comment by brandtbucher** [comment]
Thanks for taking on this project... it ended up being quite a bit more work for you than I originally thought!

I'd like to benchmark this and peek around at the generated code to make sure we don't see any significant performance degradations. I'm expecting a small slowdown on `aarch64-apple-darwin` since we need to insert trampolines there now... but we'll see how bad it really is.

> It looks like `test_traceback` and `test_tracemalloc` are failing in aarch64 linux jobs. However, those also seem to be failing in other JIT CI runs as well?

Yup, see GH-118275

File: .github/workflows/jit.yml [hunk 1]
```diff
@@ -94,35 +94,40 @@ jobs:
       - name: Native Windows
         if: runner.os == 'Windows' && matrix.architecture != 'ARM64'
         run: |
-          choco install llvm --allow-downgrade --no-progress --version ${{ matrix.llvm }}
+          choco upgrade llvm -y
+          choco install llvm --allow-downgrade --no-progress --version ${{ matrix.llvm }} --debug --verbose --noop
```
Comment (brandtbucher) [inline .github/workflows/jit.yml]: Mind explaining the addition of these options?

```suggestion
          choco install llvm --allow-downgrade --no-progress --version ${{ matrix.llvm }}
```

File: .github/workflows/jit.yml [hunk 2]
```diff
@@ -94,35 +94,40 @@ jobs:
...
+          choco upgrade llvm -y
+          choco install llvm --allow-downgrade --no-progress --version ${{ matrix.llvm }} --debug --verbose --noop 
           ./PCbuild/build.bat --experimental-jit ${{ matrix.debug && '-d' || '--pgo' }} -p ${{ matrix.architecture }}
           ./PCbuild/rt.bat ${{ matrix.debug && '-d' }} -p ${{ matrix.architecture }} -q --exclude ${{ matrix.exclude }} --multiprocess 0 --timeout 4500 --verbose2 --verbose3
 
       # No PGO or tests (yet):
       - name: Emulated Windows
         if: runner.os == 'Windows' && matrix.architecture == 'ARM64'
         run: |
+          choco upgrade llvm -y
```
Comment (brandtbucher) [inline .github/workflows/jit.yml:106]: Is this similar in spirit to `brew update`?

File: Python/jit.c [hunk 3]
```diff
@@ -237,6 +237,7 @@ patch(unsigned char *base, const Stencil *stencil, uintptr_t patches[])
             case HoleKind_IMAGE_REL_ARM64_BRANCH26:
             case HoleKind_R_AARCH64_CALL26:
             case HoleKind_R_AARCH64_JUMP26:
+            case HoleKind_ARM64_RELOC_BRANCH26:
```
Comment (brandtbucher) [inline Python/jit.c]: Nit: can you alphabetize this subgroup, like the others?

File: Tools/jit/README.md [hunk 4]
```diff
@@ -23,17 +23,21 @@ sudo ./llvm.sh 16
...
 Homebrew won't add any of the tools to your `$PATH`. That's okay; the build script knows how to find them.
 
 ### Windows
 
-Install LLVM 16 [by searching for it on LLVM's GitHub releases page](https://github.com/llvm/llvm-project/releases?q=16), clicking on "Assets", downloading the appropriate Windows installer for your platform (likely the file ending with `-win64.exe`), and running it. **When installing, be sure to select the option labeled "Add LLVM to the system PATH".**
+Install LLVM 18 [by searching for it on LLVM's GitHub releases page](https://github.com/llvm/llvm-project/releases?q=18), clicking on "Assets", downloading the appropriate Windows installer for your platform (likely the file ending with `-win64.exe`), and running it. **When installing, be sure to select the option labeled "Add LLVM to the system PATH".**
+
+## Devcontainers
+
+If you are working CPython in a [Codespaces instance](https://devguide.python.org/getting-started/setup-building/#using-codespaces), there's no need to install LLVM as the Fedora 40 base image includes LLVM 18 out of the box.
```
Comment (brandtbucher) [inline Tools/jit/README.md]: This is super cool.

File: Tools/jit/_schema.py [hunk 5]
```diff
@@ -89,6 +92,10 @@ class COFFSection(typing.TypedDict):
     SectionData: typing.NotRequired[dict[typing.Literal["Bytes"], list[int]]]
     Symbols: list[dict[typing.Literal["Symbol"], _COFFSymbol]]
 
+class _ElfSectionType(typing.TypedDict):
+    """An ELF object file section type."""
+    Name: str
+    Value: int
```
Comment (brandtbucher) [inline Tools/jit/_schema.py]: Are we actually using the `Value` field here? If not, we can remove this and just use `dict[typing.Literal["Name"], str]` inline below.

File: Tools/jit/_stencils.py [hunk 6]
```diff
@@ -203,7 +203,7 @@ def process_relocations(self, *, alignment: int = 1) -> None:
         """Fix up all GOT and internal relocations for this stencil group."""
         for hole in self.code.holes.copy():
             if (
-                hole.kind in {"R_AARCH64_CALL26", "R_AARCH64_JUMP26"}
+                hole.kind in {"R_AARCH64_CALL26", "R_AARCH64_JUMP26", "ARM64_RELOC_BRANCH26"}
```
Comment (brandtbucher) [inline Tools/jit/_stencils.py]: I try to keep `black` code style in these files. This line's getting a bit long:

```suggestion
                hole.kind in {
                    "R_AARCH64_CALL26", "R_AARCH64_JUMP26", "ARM64_RELOC_BRANCH26"
                }
```

**Comment by brandtbucher** [comment]
Also, thanks for digging up the related LLVM issues!

File: .github/workflows/jit.yml (hunk 2 above)
Comment (savannahostrowski) [reply .github/workflows/jit.yml:106]: Yep, similar! See https://docs.chocolatey.org/en-us/choco/commands/upgrade for more info!

**Comment by savannahostrowski** [comment]
I think we might be missing some tests or checks for `x86_64 Linux` jobs. The CI is passing but when I compile in a Codespace (not local dev container...which is Arm in my case), I get other unhandled instructions (e.g. `HoleKind_R_X86_64_32`, `HoleKind_R_X86_64_32S` and `HoleKind_R_X86_64_PLT32`). I'll dig into this over the weekend and patch things up but we may want to investigate adding more coverage here.

**Comment by brandtbucher** [comment]
> I get other unhandled instructions (e.g. `HoleKind_R_X86_64_32`, `HoleKind_R_X86_64_32S` and `HoleKind_R_X86_64_PLT32`).

Hm, I wonder if Fedora uses a patched LLVM with different defaults or something...

**Comment by brandtbucher** [comment]
Can you try adding `-fpic` to the list of `args` for the Linux targets? Position-independent code (which lets us reference symbols more than 32 bits away) is the default on the other platforms, but based on your description it seems like it isn't for Fedora. Adding that option might fix it.

**Comment by savannahostrowski** [comment]
Ah! Okay, that did work for the Codespace. Also, checked in the local dev container as well. Thanks for the tip!

**Comment by brandtbucher** [comment]
Nice, looks like it's also [a slight performance improvement](https://github.com/faster-cpython/benchmarking-public/blob/main/results/bm-20240426-3.13.0a6%2B-fe17f68-JIT/bm-20240426-linux-x86_64-savannahostrowski-llvm_18-3.13.0a6%2B-fe17f68-vs-base.png) (if anything).

**Comment by brandtbucher** [comment]
Doc failure isn't us: https://github.com/python/cpython/issues/118401

**Comment by brandtbucher** [comment]
Free-threading failures look like flukes too. Gonna re-run once the JIT tests finish.

**Pull request merged**